File: cpp/2415.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
private:
    void reverse(TreeNode* r1, TreeNode* r2, bool odd) {
        if (r1 == NULL)
            return;
        if (odd) {
            int t = r1->val;
            r1->val = r2->val;
            r2->val = t;
        }
        reverse(r1->left, r2->right, !odd);
        reverse(r1->right, r2->left, !odd);
    }

public:
    TreeNode* reverseOddLevels(TreeNode* root) {
        reverse(root->left, root->right, true);
        return root;
    }
};

File: python/2415.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def reverseOddLevels(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def reverse(r1: Optional[TreeNode], r2: Optional[TreeNode], odd: bool):
            if r1 is None:
                return
            if odd:
                r1.val, r2.val = r2.val, r1.val
            reverse(r1.left, r2.right, not odd)
            reverse(r1.right, r2.left, not odd)

        reverse(root.left, root.right, True)
        return root
